File: reorderDataInLogFiles/c++/reorder_data_in_log_files.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

class Solution {
public:
    vector<string> reorderLogFiles(vector<string>& logs) {
        vector<string> result;
        int count = 0;

        for (auto log: logs){
            if (log.back() <= 'z' and log.back() >= 'a') {
                result.insert(result.begin(), log);
                count ++;
            }
            else{
                result.push_back(log);
            }
        }
        sort(result.begin(), result.begin() + count, compare);
        return result;
    }
    static bool compare(string s1, string s2) {
        string a = s1.substr(s1.find(' '));
        string b = s2.substr(s2.find(' '));
        return a == b ? s1 < s2 : a < b;
    }
};

int main() {
    vector<string> logs;

    logs.insert(logs.end(), "dig1 8 1 5 1");
    logs.insert(logs.end(), "let1 art can");
    logs.insert(logs.end(), "dig2 3 6");
    logs.insert(logs.end(), "let2 own kit dig");
    logs.insert(logs.end(), "let3 art zero");

    Solution obj;
    vector<string> answer = obj.reorderLogFiles(logs);
    for (auto log: answer)
        cout << log << endl;
}

File: problems/reorderDataInLogFiles/python/reorder_data_in_log_files.py
from typing import List


class Solution:
    """
    Complexity
    ----------
    - M: maximum length of a single list
    - N: # of logs in the list
    - TC: O(M Nlog(N))
        - O(N log(N)): `sorted` in python is based on Timsort algorithm
        - Since the keys of the elements are basically the logs itself, the comparison between two keys can take up to O(M) time.
    - SC: O(MN)
        - O(MN): `sorted` in python is based on Timsort algorithm
    """

    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        def get_key(log):
            identifier, rest_log = log.split(" ", maxsplit=1)
            # (indicator of the type of logs, content, id)
            # indicator of the type of logs = 0: letter log, 1: digit log
            return (0, rest_log, identifier) if rest_log[0].isalpha() else (1,)

        return sorted(logs, key=get_key)


if __name__ == "__main__":
    logs = [
        "dig1 8 1 5 1",
        "let1 art can",
        "dig2 3 6",
        "let2 own kit dig",
        "let3 art zero",
    ]
    print(f"The solution is {Solution().reorderLogFiles(logs)}")
